### ISSUE TITLE
Feature Request: Add separable_conv2d_transpose operation

### ISSUE THREAD

**Issue by andreas-eberle** [tensorflow/tensorflow#12001]
Some recent papers (e.g.) have shown that transposed separable convolutions can be a great choice for decoders in encoder decoder architectures.

Can you add a seperable_conv2d_transpose operation comparable to the conv2d_transpose operation?

**Comment by tjingrant** [comment]
Hi, I will try to work on this one.

@andreas-eberle do you have specific examples in which a seperable_conv2d_transpose operation is profitable?

**Comment by expelliarms** [comment]
Please provide the links of the reference papers

**Comment by andreas-eberle** [comment]
Sorry, didn't notice that I forgot the link. 

In [The Devil is in the Decoder](https://arxiv.org/abs/1707.05847) they compare several "deconvolution" strategies and show that separable transposed convolution has very good performance.

In section 2.1.1 they give a short explanation about separable transposed convolution.

Afaik, separable (not transposed) convolution was introduced in [Xception: Deep Learning with Depthwise Separable Convolutions](https://arxiv.org/abs/1610.02357)

**Comment by ezfn** [comment]
+1 on that. 
Besides the examples given by andreas-eberle, It is generally helpful to have operations that are not allowed to mix the channels, so that channels from different layers can be trivially combined (i.e. summed without any learned parameters).

**Comment by titusnicolae-intel** [comment]
Hi, is anyone working on this anymore?
I'd like to start work on this, if anyone can provide some supervision that would be helpful.

**Comment by estathop** [comment]
useful feature to be implemented

**Comment by dhaneshr** [comment]
any updates on this ?

**Comment by notnot** [comment]
I'd like to try a GAN with separable_conv2d and separable_conv2d_transpose, and i was surprised to see that separable_conv2d_transpose isn't available yet. Some have stated they started working on an implementation, how is that work going?

**Comment by chris-boson** [comment]
I would also be able to help with an implementation, @tjingrant have you started on it? 
Also seems like the authors of [The Devil is in the Decoder](https://arxiv.org/abs/1707.05847) must have access to it.

**Comment by HouseOfFinwe** [comment]
Any idea when there will be an implementation for seaparable_conv2d_tranpose?

**Comment by brucechou1983** [comment]
Any update to this feature?

**Comment by chris-boson** [comment]
Easiest to just use`tf.keras.layers.Conv2DTranspose` followed by `tf.keras.layers.DepthwiseConv2D`

**Comment by HouseOfFinwe** [comment]
@chris-boson. This is not equivalent.  One of the major selling points of DepthwiseConv2DTranspose (if it existed) is a reduction of parameters, which would not be achieved by a transpose followed by a depthwise conv.

**Comment by chris-boson** [comment]
@HouseOfFinwe It does in fact reduce the parameter count considerably

**Comment by netanel-s** [comment]
+1 on this, would be highly appreciated.

**Comment by veqtor** [comment]
@chris-boson could you give a clearer example of using Conv2DTranspose followed by DepthwiseConv2D?

**Comment by zhcm** [comment]
Any update to this feature?

**Comment by shenyi0220** [comment]
Any updates or progress on this?

**Comment by mbuckler** [comment]
I would be interested in this

**Comment by ltrottier** [comment]
That would be a nice addition indeed.

**Comment by ygoncharov** [comment]
Seems like a feature that makes a lot of sense

**Comment by veqtor** [comment]
Would this require a new op? Is it difficult because of lack of hw support?

**Comment by mjmjmtl-pony** [comment]
Any update on this?

**Comment by voletiv** [comment]
Any update on this?

**Comment by CoachRDeveloper** [comment]
Would like to see this feature implemented

**Comment by wmcnally** [comment]
What would be the difference between this op and Conv2DTranspose(filters=input_channels) followed by a pointwise conv?

**Comment by wmcnally** [comment]
What would be the difference between this op and Conv2DTranspose(filters=1) followed by a pointwise conv?

**Comment by edmondja** [comment]
+1

**Comment by gurpreet-singh135** [comment]
Hi, is anyone working on this currently?
I'd like to work on this feature

**Comment by edmondja** [comment]
Is it very different to use upsampling + separableconv ?

**Comment by gurpreet-singh135** [comment]
@edmondja the problem with upsampling + separableconv is that it increases the number of computations compared to sep-conv2d-transpose

**Comment by yaoshiang** [comment]
Anyone find a workaround? I attempted to do a workaround with a stack of Conv2DTranspose, each with filters=1... but it was not very efficient. No promise this works but this was my attempt fwiw. 


class DepthwiseConv2DTranspose(layers.Layer):
    def __init__(self, filters, **kwargs):
        super(DepthwiseConv2DTranspose, self).__init__(**kwargs)
        self._filters = filters
        self._t = []
        for _ in range(filters):
            self._t.append(layers.Conv2DTranspose(filters=1, kernel_size=5, strides=2, output_padding=1))

    def __call__(self, img):
        upsample = []
        for i in range(self._filters):
            t = self._t[i](img[:,:,:,i:i+1])
            t = t[:,:,:,0]
            upsample.append(t)
        upsample = tf.stack(upsample, axis=-1)
        return upsample

**Comment by Orpheus23** [comment]
Here are two sample ones that do some of it, the first doesn't compile on TPU it does work on CPU 
 ```
class Depthwise_Conv2D_Transpose(tf.keras.layers.Layer):
      def __init__(self, filters,kernel_size,strides,padding='same',use_bias=False,kernel_initializer=None,name="",**kwargs):
          super(Depthwise_Conv2D_Transpose, self).__init__(**kwargs)
          self.kernel_size = kernel_size
          self.strides = strides[0]
          self.padding = padding
          self.use_bias = use_bias
          self.kernel_init = kernel_initializer
          self.lambdas =[]
          for i in tf.range(filters):  
             self.lambdas.append(layers.Conv2DTranspose(filters=1, kernel_size=kernel_size,strides=self.strides,padding=self.padding))
          self.filters = filters
          self.input_image_shape = 0
          self.nm = name
      def call(self, inputs):
          #tf.print(inputs.shape,[-1]+[self.deconv_length(self.input_image_shape,self.strides,self.kernel_size,self.padding)]*2+[1])
          inputs_channel_wise =   tf.split(inputs,self.filters, -1)#
          x_outputs = [c(x) for x, c in zip(inputs_channel_wise, self.lambdas)]
          
          channel_wise_conv =  tf.concat(x_outputs, -1)#tf.transpose(tf.squeeze(channel_wise_conv,axis = -1),[0,2,3,1])
          return channel_wise_conv
```
 The second compiles also on TPU and works well on CPU but doesn't train on TPU
```
class Depthwise_Conv2D_Transpose(tf.keras.layers.Layer):
    def __init__(self, filters,kernel_size,strides,padding='same',use_bias=False,kernel_initializer=None,name="",**kwargs):
        super(Depthwise_Conv2D_Transpose, self).__init__(**kwargs)
        self.kernel_size = kernel_size
        self.strides = strides[0]
        self.padding = padding
        self.use_bias = use_bias
        self.kernel_init = kernel_initializer
        self.lambdas =[]
        self.filters = filters
        self.input_image_shape = 0
        self.nm = name
        
    def deconv_length(self,dim_size, stride_size, kernel_size, padding, output_padding=None, dilation=1):

        assert padding in {'same', 'valid', 'full'}
        if dim_size is None:
            return None

        # Get the dilated kernel size
        kernel_size = kernel_size + (kernel_size - 1) * (dilation - 1)

        # Infer length if output padding is None, else compute the exact length
        if output_padding is None:
            if padding == 'valid':
                dim_size = dim_size * stride_size + max(kernel_size - stride_size, 0)
            elif padding == 'full':
                dim_size = dim_size * stride_size - (stride_size + kernel_size - 2)
            elif padding == 'same':
                dim_size = dim_size * stride_size
        else:
            if padding == 'same':
                pad = kernel_size // 2
            elif padding == 'valid':
                pad = 0
            elif padding == 'full':
                pad = kernel_size - 1

            dim_size = ((dim_size - 1) * stride_size + kernel_size - 2 * pad + output_padding)

        return dim_size

    def build(self, input_shape):
        for i in range(input_shape[-1]):  
           self.lambdas.append(self.add_weight(name = self.nm +"weights"+ str(i),initializer=tf.keras.initializers.deserialize(self.kernel_init),shape=(self.kernel_size,self.kernel_size,1,1), trainable=True))
        self.input_image_shape = input_shape[1]
        #self.lambdas = tf.stack(self.lambdas,axis = 0)
        self.image_shape = input_shape[-1]
        super(Depthwise_Conv2D_Transpose, self).build(input_shape)
    @tf.function
    def call(self, inputs):
        
        inputs_channel_wise =   tf.split(inputs,self.image_shape, -1)

        channel_wise_conv = tf.map_fn(lambda x:tf.nn.conv2d_transpose(x[0], filters=x[1], 
                                                                              output_shape=[tf.shape(inputs)[0]]+[self.deconv_length(self.input_image_shape,self.strides,self.kernel_size,self.padding)]*2+[1], 
                                                                              strides=self.strides, 
                                                                              padding=self.padding.upper()),(inputs_channel_wise,self.lambdas), fn_output_signature=tf.float32)
        
        channel_wise_conv = tf.transpose(tf.squeeze(channel_wise_conv,axis = -1),[0,2,3,1])
        return channel_wise_conv  
```      
If any solutions found do update thanks!

**Comment by Ram-WD** [comment]
any update ?

**Comment by glenn-jocher** [comment]
@Orpheus23 is your TF Depthwise Conv2dTranspose implementation in https://github.com/tensorflow/tensorflow/issues/12001#issuecomment-822746500 still the best available today?

**Comment by Orpheus23** [comment]
Idk about any recent changes, but until early 2021(When I was checking) it was. If a better solution is needed, then it is best to write the same in c++ and export it. The layers that I had written didn't train properly on TPU(Idk about GPU, if it worked for someone then do update) as they required a lot of space in memory. Perhaps it is because of the mapping of conv2d transpose to every layer and the space required by tf.transpose + tf.squeeze. 

Either way it would best to define on c++, the functions that tensorflow provides on python are very unideal for defining a new type of layer. With c++ there could be a better alternative to mapping conv2d_transpose to every layer and then reshaping everything.

**Comment by glenn-jocher** [comment]
@Orpheus23 hmm, yes that's what I was worried about. We've been running YOLOv5 experiments with these layers in PyTorch, and they seem to export well everywhere except TF. We build the TF models natively rather than go through ONNX, and currently there seems to be no efficient solution to build TF models with these layers. In PyTorch it's pretty simple as you can just set groups to equal the input/output channel counts to create depthwise conv2dtranspose layers.

**Comment by Orpheus23** [comment]
That's there, Pytorch works very well in these matters. The group feature is not here, maybe with tf.gather and conv2d transpose you might be able to do it, it will not be very clean. If you really want to go ahead with TF here, then C++ is the best bet. Integrating the C++ part with python was not that much of a headache. Otherwise the other alternatives are, to do bi-linear interpolation and then depth-wise conv2d or to hack it out with TF in python. In any case, Best of luck with the implementation.

**Comment by glenn-jocher** [comment]
@Orpheus23 got it, thanks! I'm linking @zldrobit here, who's been a lot of the brains behind the YOLOv5 TF models.

**Comment by glenn-jocher** [comment]
@ayushexel @sergiossm this is the main issue I found regarding DW Conv2d Transpose layers in TF (C++ conversion proposed by @Orpheus23)

**Comment by saikrn112** [comment]
Hey guys, any update on this feature?

**Comment by bayesian-mind** [comment]
Any update on the feature request for depthwise conv2d transpose?

**Comment by sushreebarsa** [comment]
Hi,

Could you set the groups parameter to the number of input/output channels in a tf.nn.separable_conv2d layer followed by a tf.nn.conv2d_transpose layer.

Since this issue has been open for a long time, the code/debug information for this issue may not be relevant with the current state of the code base.

The Tensorflow team is constantly improving the framework by fixing bugs and adding new features. We suggest you try the latest TensorFlow version with the latest compatible hardware configuration which could potentially resolve the issue. If you are still facing the issue, please create a new GitHub issue with your latest findings, with all the debugging information which could help us investigate.

Please follow the release notes to stay up to date with the latest developments which are happening in the Tensorflow space.

Thank you!

**Comment by eewindfly** [comment]
Any update on this feature?
Pytorch can do this in very simple way, just correctly config the groups parameter.
I really hope tensorflow can do this. 
Considering to current situation, which SeparableConv2D has existed, open a new function SeparableConv2DTranspose is fine too.
